### PR TITLE
Add support for AD7124

### PR DESCRIPTION
**Pull request by adisuciu**
The AD7124 driver still has some issues, however this should be the API.

A few notes:
- Currently it only works when one channel is enabled.
- There are sample rates for all channels (and they are not shared attributes), however i'm not sure how it is supposed to work when having multiple channels enabled - same buffer, multiple rates. This is why I set the sample rate for all channels
- The scale and scale_available attribute was trickier because the str(float val) gives a value like 4.56+e06 which is incompatible with IIO.
- I'm not sure where the to_volts method should be implemented as it uses different coeficients for different channels.
- I created the channels list on __init__ dynamically. I do this because the number of channels is reconfigurable via devicetree. Does it make sense reffering to channels by index ? Or you should refer to channels by name. A channel name for ad7124 is "input voltageX-voltageY" because the input is always differential.
- The example just gets data from the ADC. I can create some plots when the driver will work properly. It currently gives out junk values.

Signed-off-by: Adrian Suciu <adrian.suciu@analog.com>

### CODE REVIEW

File: adi/ad7124.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    @property
+    def sampling_frequency(self):
+        # only works with the same frequency for all channels
+        return self._get_iio_attr(self.channel[0].name, "sampling_frequency", False)
+
+    @sampling_frequency.setter
+    def sampling_frequency(self, value):
+        for ch in self.channel:
+            self._set_iio_attr(ch.name, "sampling_frequency", False, value)
+        return 1
```
Comment (tfcollins) [inline adi/ad7124.py]: Decorator based setters don't return things in python

File: adi/ad7124.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    _device_name = ""
+
+    def __init__(self, uri=""):
+
+        context_manager.__init__(self, uri, self._device_name)
+        self._ctrl = self._ctx.find_device("ad7124-8")
+        self._rxadc = self._ctx.find_device("ad7124-8")
+
+        # dynamically get channels
+        for ch in self._ctrl._channels:
```
Comment (tfcollins) [inline adi/ad7124.py]: I assume this works since all channels have scan elements?

File: adi/ad7124.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    def sampling_frequency(self, value):
+        for ch in self.channel:
+            self._set_iio_attr(ch.name, "sampling_frequency", False, value)
+        return 1
+
+    @property
+    def scale_available(self):
+        return self._get_iio_attr(self.channel[0].name, "scale_available", False)
+
+    class _channel(attribute):
```
Comment (tfcollins) [inline adi/ad7124.py]: This is pretty handy to have. Do you think we could possibly combine this into the rx_tx class so data received from buffers are automatically converted to specific units? Say we had a property like rx_output_type = 'SI' or something?

The adis16460 class would benefit.

File: adi/ad7124.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        # dynamically get channels
+        for ch in self._ctrl._channels:
+            name = ch._id
+            self._rx_channel_names.append(name)
+            self.channel.append(self._channel(self._ctrl, name))
+        rx.__init__(self)
+
+    @property
+    def sampling_frequency(self):
+        # only works with the same frequency for all channels
```
Comment (tfcollins) [inline adi/ad7124.py]: Can you add docstrings to each property

File: examples/ad7124.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+from scipy import signal
+
+# Set up AD7124
+ad7124 = adi.ad7124(uri="ip:192.168.1.171")
+
+sc = ad7124.scale_available
+ad7124.channel[0].scale = sc[2]
+
+ad7124.sampling_frequency = 19200 # sets sample rate for all channels
+ad7124.rx_enabled_channels = [ 0 ] # currently only one enabled channel buffer works at a time
```
Comment (tfcollins) [inline examples/ad7124.py]: Is this how to it works with IIO-Scope or maybe something is wrong with core pyadi-iio which doesn't support this case?

**Comment by tfcollins** [comment]
One reason you might be getting bad data is that the buffers default to np.int16 samples and the AD7123 is a 24-bit device (I believe) which are probably put into a 32-bit container in DDR. You can update the expected type like this: https://github.com/analogdevicesinc/pyadi-iio/blob/master/adi/adis16460.py#L52

File: adi/ad7124.py (hunk 2 above)
Comment (adisuciu) [reply adi/ad7124.py]: Yes. Is there any other way to do it to catch channels without buffers too ?

File: examples/ad7124.py (hunk 5 above)
Comment (adisuciu) [reply examples/ad7124.py]: It's the same in IIO-oscilloscope. It looks like it's a driver thing.

File: adi/ad7124.py (hunk 1 above)
Comment (adisuciu) [reply adi/ad7124.py]: added

File: adi/ad7124.py (hunk 4 above)
Comment (adisuciu) [reply adi/ad7124.py]: added

**Comment by adisuciu** [comment]
> One reason you might be getting bad data is that the buffers default to np.int16 samples and the AD7123 is a 24-bit device (I believe) which are probably put into a 32-bit container in DDR. You can update the expected type like this: https://github.com/analogdevicesinc/pyadi-iio/blob/master/adi/adis16460.py#L52

I changed the return type. Some of the data is valid because there is a correlation between the test signal and the plot. I get data interleaved with zeroes and some junk data as well. It's the same in IIO-Oscilloscope.

File: adi/ad7124.py (hunk 3 above)
Comment (adisuciu) [reply adi/ad7124.py]: Do you think this should be a separate pull request ?

Comment (adisuciu) [reply adi/ad7124.py]: I added another commit that handles _rx_output_type .. seems to work for AD7124. Does the implementation make sense. I don't have a ADIS16460. I can probably test it tomorrow if I find one in the office.

Comment (tfcollins) [reply adi/ad7124.py]: There is one inside the PackRF

File: adi/ad7124.py (hunk 2 above)
Comment (tfcollins) [reply adi/ad7124.py]: You could check if the scan_element properties of the channel objects are true and if they are inputs.

File: examples/ad7124.py (hunk 5 above)
Comment (tfcollins) [reply examples/ad7124.py]: I guess we'll live with it for now. Will revisit when the driver is done/fixed

**Pull request merged**